j-r/minetest
# WIP: optional new liquid behavior

This is a test PR to solicit feedback concerning the usefulness of the new liquid behavior implemented in this branch, primarily from the Mineclonia community.

It differs from other attempts in that it remains fully within the framework of the current minetest liquid implementation and therefore should be bug-for-bug compatible with current minetest unless the new liquid behavior is explicitly switched on. It adds two new param types, ``directionalflowing`` and ``directionalsource`` with the following changes in behavior

- flowing in a single direction if there is a path within liquid_directed_range to a lower level
- when two or more different liquids meet in a node the strongest flow will displace the other liquids
- a liquid source node that flows down will not spread horizontally
- flowing liquid above a source node of the same kind will not spread horizontally
- rendering is slightly different; older clients will render linear flows without volume, but otherwise work fine

As a special bonus feature directional liquid node tops are rendered 1/16 node lower. But unlike most of the rest of the series, that is just an ugly hack. 

Note that this branch is currently based on a rather old version of minetest 5.9-dev, because that is the version I've been using for some time on my private server and am pretty sure that it has no issues interfering with testing these liquid physics changes.

To test it out

1. check out the branch and build it normally in your environment. (Note that while most of the changes are compiled into the ``minetest`` and ``minetestserver`` executables, ``builtin/game/register.lua`` and ``builtin/settingtypes.txt`` also have relevant changes.) 

2. run it from the build directory or from a suitable installation directory

3. create a new world or take a copy of an existing world (it should be non destructive in the sense that no liquid should flow where it didn't flow before, but there may be bugs)

4. go to the settings and switch the ``force_directional_liquids`` setting

5. run your world

6. on page 3 of the F6 engine profiler you should see some statistics about liquid processing; when creating a new world or loading a world that didn't have directional liquids, there will be a lot of activity on initial loading; it should become stable after a while and only do anything when loading new blocks (e.g. when turning and moving): stopping and reloading the world should have very little activity

7. check out how the water behaves now

8. check out the ``force_liquid_directed_range`` setting that will limit the length of linear flows

9. yell at me that it's crap:-)

The primary goal is to get feedback on whether this is actually useful, what is good, what is bad/missing etc.So please do that. Thank you very much!

Note that this isn't MC liquid even if the results may in some cases look like MC, but it still is very much the minetest liquid system and has different quirks, that might not be fixable within the given framework. The benefit of doing it this way *may* be that it avoids some of the pitfalls that prevented other attempts from getting accepted.
